firmware/development/perf_trace.cpp

```diff
@@ -0,0 +1,93 @@
+
+#include "efifeatures.h"
+#include "perf_trace.h"
+#include "efitime.h"
+#include "os_util.h"
+
+#define TRACE_BUFFER_LENGTH 2048
+
+enum class EPhase : char
+{
+	Start,
+	End,
+	InstantThread,
+	InstantGlobal,
+};
+
+struct TraceEntry
+{
+	PE Event;
+	EPhase Phase;
+	uint8_t Data;
+	uint8_t ThreadId;
+	uint32_t Timestamp;
+};
+
+static_assert(sizeof(TraceEntry) == 8);
+
+static TraceEntry s_traceBuffer[TRACE_BUFFER_LENGTH];
```

Comment:
We still need  #if ENABLE_PERF_TRACE because of large s_traceBuffer.

Reply:
yes - I haven't guarded the stuff in here yet with `ENABLE_PERF_TRACE`